src/algebra/hom/centroid.lean

```diff
@@ -0,0 +1,289 @@
+    .. (n • f : α →+ α) }⟩
+
+instance has_npow_nat : has_pow (centroid_hom α) ℕ :=
+⟨λ f n,
+{  map_mul_left' := λ a b, begin
+    induction n with n ih,
+    { simp },
+    { rw pow_succ,
+      exact (congr_arg f.to_End ih).trans (f.map_mul_left' _ _) }
+  end,
+  map_mul_right' := λ a b, begin
+    induction n with n ih,
+    { simp },
+    { rw pow_succ,
+      exact (congr_arg f.to_End ih).trans (f.map_mul_right' _ _) }
+  end,
+  ..(f.to_End ^ n : add_monoid.End α) }⟩
+
+@[simp] lemma coe_zero : ⇑(0 : centroid_hom α) = 0 := rfl
+@[simp] lemma coe_one : ⇑(1 : centroid_hom α) = id := rfl
+@[simp] lemma coe_add (f g : centroid_hom α) : ⇑(f + g) = f + g := rfl
+@[simp] lemma coe_mul (f g : centroid_hom α) : ⇑(f * g) = f ∘ g := rfl
+-- Eligible for `dsimp`
+@[simp, nolint simp_nf] lemma coe_nsmul (f : centroid_hom α) (n : ℕ) : ⇑(n • f) = n • f := rfl
+
+@[simp] lemma zero_apply (a : α) : (0 : centroid_hom α) a = 0 := rfl
+@[simp] lemma one_apply (a : α) : (1 : centroid_hom α) a = a := rfl
+@[simp] lemma add_apply (f g : centroid_hom α) (a : α) : (f + g) a = f a + g a := rfl
+@[simp] lemma mul_apply (f g : centroid_hom α) (a : α) : (f * g) a = f (g a) := rfl
+-- Eligible for `dsimp`
+@[simp, nolint simp_nf]
+lemma nsmul_apply (f : centroid_hom α) (n : ℕ) (a : α) : (n • f) a = n • f a := rfl
+
+-- cf.`add_monoid_hom.add_comm_monoid`
+instance : add_comm_monoid (centroid_hom α) :=
+coe_to_add_monoid_hom_injective.add_comm_monoid _ rfl (λ _ _, rfl) (λ _ _, rfl)
+
+instance : add_comm_monoid_with_one (centroid_hom α) :=
+{ ..centroid_hom.add_comm_monoid,
+  ..centroid_hom.has_one }
+
+-- cf `add_monoid.End.semiring`
+instance : semiring (centroid_hom α) :=
+to_End_injective.semiring _ (by { ext, refl }) (by { ext, refl }) (λ _ _, rfl) (λ _ _, rfl)
+  (λ _ _, rfl) (λ _ _, by { ext, refl }) (λ n, begin
+    ext,
+    unfold to_End,
+    induction n,
+    rw [nat.cast_zero, coe_to_add_monoid_hom, zero_apply, nat.cast_zero, add_monoid_hom.zero_apply],
+    rw [nat.cast_succ, nat.cast_succ, add_monoid_hom.add_apply, ← n_ih, coe_to_add_monoid_hom,
+      add_apply, one_apply, coe_to_add_monoid_hom, add_monoid.coe_one, id.def],
+  end)
+
+lemma comp_mul_comm (T S : centroid_hom α) (a b : α) : (T ∘ S) (a * b) = (S ∘ T) (a * b) :=
+by rw [comp_app, map_mul_right, map_mul_left, ←map_mul_right, ←map_mul_left]
+
+end non_unital_non_assoc_semiring
+
+section non_unital_non_assoc_ring
+variables [non_unital_non_assoc_ring α]
+
+/-- Negation of `centroid_hom`s as a `centroid_hom`. -/
+instance : has_neg (centroid_hom α) :=
+⟨λ f,
+  { map_mul_left' := by simp [map_mul_left],
+    map_mul_right' := by simp [map_mul_right],
+    .. (-f : α →+ α) }⟩
+
+
+instance : has_sub (centroid_hom α) :=
+⟨λ f g,
+{ map_mul_left' := λ a b, by simp [map_mul_left, mul_sub],
+  map_mul_right' := λ a b, by simp [map_mul_right, sub_mul],
+  .. (f - g : α →+ α) }⟩
+
+instance has_scalar_int : has_smul ℤ (centroid_hom α) :=
```

Comment:
All these `has_scalar`s should be renamed to `has_smul` to match the typeclass being renamed.